File: golang/binary-tree-dfs/longest_zigzag_binary_tree.go
package main

func getLeafSequence(root *TreeNode, leaves *[]int) {
	if root == nil {
		return
	}

	if root.Left == nil && root.Right == nil {
		*leaves = append(*leaves, root.Val)
	}

	getLeafSequence(root.Left, leaves)
	getLeafSequence(root.Right, leaves)
}

func leafSimilar(root1 *TreeNode, root2 *TreeNode) bool {
	var leaves1, leaves2 []int

	getLeafSequence(root1, &leaves1)
	getLeafSequence(root2, &leaves2)

	if len(leaves1) != len(leaves2) {
		return false
	}

	for i := 0; i < len(leaves1); i++ {
		if leaves1[i] != leaves2[i] {
			return false
		}
	}

	return true
}

File: java/src/main/java/com/sanedge/interview/binary_tree_dfs/LongestZigzagBinaryTree.java
package com.sanedge.interview.binary_tree_dfs;

import com.sanedge.interview.utils.TreeNode;
import java.util.ArrayList;
import java.util.List;

public class LongestZigzagBinaryTree {

    public void getLeafSequence(TreeNode root, List<Integer> leaves) {
        if (root == null) {
            return;
        }

        if (root.left == null && root.right == null) {
            leaves.add(root.val);
        }

        getLeafSequence(root.left, leaves);
        getLeafSequence(root.right, leaves);
    }

    public boolean leafSimilar(TreeNode root1, TreeNode root2) {
        List<Integer> leaves1 = new ArrayList<>();
        List<Integer> leaves2 = new ArrayList<>();

        getLeafSequence(root1, leaves1);
        getLeafSequence(root2, leaves2);

        if (leaves1.size() != leaves2.size()) {
            return false;
        }

        for (int i = 0; i < leaves1.size(); i++) {
            if (!leaves1.get(i).equals(leaves2.get(i))) {
                return false;
            }
        }

        return true;
    }
}
